### ISSUE TITLE
Rethink cache header strategy

### ISSUE THREAD

**Issue by rexxars** [imbo/imbo#395]
This is related to #267.

I think the cache headers we currently use for endpoints like `/images` and `/metadata` is wrong.

Currently, a request to `/images` results in something like this:

```
Cache-Control: public
ETag: "e16a1a3a6ca89f0bbee14b7ba6583f97"
Last-Modified: Thu, 05 Nov 2015 13:57:34 GMT
```

I'm having a bit of a hard time understanding how browsers implement their caching strategy, but it seems obvious to me at this point that the results are currently pretty unpredictable.

After talking to a number of people, I think the best we can do (to ensure that browsers get the latest response, always) is to send `Cache-Control: max-age=0, must-revalidate, public` (probably configurable), and implement some sort of configurable http cache listener.

This listener could for instance send:
```
Surrogate-Control: max-age=7200
Surrogate-Key: some-unique-key-for-this-endpoint
```

Given a configured http cache adapter, this can trigger a purge for external caches (varnish, cdns etc) when something happens that causes the endpoint to be invalidated.

**Comment by rexxars** [comment]
Instagram and Flickr both serve their clients cache headers which basically all say "don't cache", if that makes any difference here. Regardless:

I've been thinking about which resources that potentially suffer from this problem, and in summary, I think it's pretty much every single one:
 * status
 * stats
 * user
 * global images
 * images
 * metadata
 * groups
 * group
 * keys
 * access rules
 * access rule

All of these serve data that could change at any time based on external factors (other API requests manipulating the data, obviously).

What I'm currently thinking of implementing:
- `httpCache` section in the configuration which allows server admins to set up their own values. Current suggestion for defaults is `max-age=0, must-revalidate, public` - open to suggestions here
- Introduce a `HttpCache`-listener that will trigger right before sending the response, on all requests. It will check if the resource has any cache-control settings applied, and if not, set the defaults configured. For instance, the `image` resource explicitly sets `max-age` to `31536000` (1 year), and the defaults would not override this.

Setting `surrogate-control` and `surrogate-key` is beyond the scope of this issues, and should be implemented as a standalone event listener.

**Comment by rexxars** [comment]
Solved in #398
